Kernel/FileSystem/Plan9FileSystem.cpp

```diff
@@ -0,0 +1,752 @@
+
+    auto tmp_buffer = m_builder.build();
+
+    m_have_been_built = true;
+    m_builder.~KBufferBuilder();
+
+    new (&m_built.buffer) KBuffer(move(tmp_buffer));
+    new (&m_built.decoder) Decoder({ m_built.buffer.data(), m_built.buffer.size() });
+    u32* size = reinterpret_cast<u32*>(m_built.buffer.data());
+    *size = m_built.buffer.size();
+    return m_built.buffer;
+}
+
+KResult Plan9FS::post_message(Message& message)
+{
+    auto& buffer = message.build();
+    const u8* data = buffer.data();
+    size_t size = buffer.size();
+    auto& description = file_description();
+
+    LOCKER(m_send_lock);
+
+    while (size > 0) {
+        if (!description.can_write()) {
+            if (Thread::current()->block<Thread::WriteBlocker>(description) != Thread::BlockResult::WokeNormally)
+                return KResult(-EINTR);
+        }
+        ssize_t nwritten = description.write(data, size);
+        if (nwritten < 0)
+            return KResult(nwritten);
+        data += nwritten;
+        size -= nwritten;
+    }
+
+    return KSuccess;
+}
+
+KResult Plan9FS::do_read(u8* data, size_t size)
+{
+    auto& description = file_description();
+    while (size > 0) {
+        if (!description.can_read()) {
+            if (Thread::current()->block<Thread::ReadBlocker>(description) != Thread::BlockResult::WokeNormally)
+                return KResult(-EINTR);
+        }
+        ssize_t nread = description.read(data, size);
+        if (nread < 0)
+            return KResult(nread);
+        if (nread == 0)
+            return KResult(-EIO);
+        data += nread;
+        size -= nread;
+    }
+    return KSuccess;
+}
+
+KResult Plan9FS::read_and_dispatch_one_message()
+{
+    ASSERT(m_someone_is_reading);
+    // That someone is us.
+
+    struct [[gnu::packed]] Header
+    {
+        u32 size;
+        u8 type;
+        u16 tag;
+    };
+    Header header;
+    KResult result = do_read(reinterpret_cast<u8*>(&header), sizeof(header));
+    if (result.is_error())
+        return result;
+
+    auto buffer = KBuffer::create_with_size(header.size, Region::Access::Read | Region::Access::Write);
+    // Copy the already read header into the buffer.
+    memcpy(buffer.data(), &header, sizeof(header));
+    result = do_read(buffer.data() + sizeof(header), header.size - sizeof(header));
+    if (result.is_error())
+        return result;
+
+    LOCKER(m_lock);
+
+    auto optional_completion = m_completions.get(header.tag);
+    if (!optional_completion.has_value()) {
+        if (m_tags_to_ignore.contains(header.tag)) {
+            m_tags_to_ignore.remove(header.tag);
+        } else {
+            dbg() << "Received a 9p message of type " << header.type << " with an unexpected tag " << header.tag << ", dropping";
+        }
+        return KSuccess;
+    }
+    ReceiveCompletion& completion = *optional_completion.value();
+    completion.result = KSuccess;
+    completion.message.~Message();
+    new (&completion.message) Message(move(buffer));
```

Comment:
Can we write this as a regular assignment, or does the placement new sidestep something here?

Reply:
No regular assignment unfortunately ☹️. The reason is, because of the `union` (and `KBuffer` which doesn't have a null state — or perhaps it does, hmm) you cannot assign or move `Message`s. With explicit destructor + placement new, we're not assigning or moving anything, so it works.

That being said, I need to actually experiment with overriding `operator=`, perhaps that would be not super ugly and clean up some code like this.

Reply:
`operator=()` actually seems to work